collect_app/src/main/java/org/odk/collect/android/logic/AuditEvent.java

```diff
@@ -119,14 +138,39 @@ public void setLocationCoordinates(String latitude, String longitude, String acc
         this.accuracy = accuracy;
     }
 
+    public void setNewValue(String newValue) {
+        this.newValue = newValue != null ? newValue : "";
+    }
+
     /*
      * convert the event into a record to write to the CSV file
      */
     @NonNull
     public String toString() {
-        return hasLocation()
-                ? String.format("%s,%s,%s,%s,%s,%s,%s", auditEventType.getValue(), node, start, end != 0 ? end : "", latitude, longitude, accuracy)
-                : String.format("%s,%s,%s,%s", auditEventType.getValue(), node, start, end != 0 ? end : "");
+        String node = formIndex == null || formIndex.getReference() == null ? "" : formIndex.getReference().toString();
+        int idx = node.lastIndexOf('[');
+        if (idx > 0) {
+            node = node.substring(0, idx);
+        }
+
+        // Clear values if they are equal
+        if (oldValue.equals(newValue)) {
```

Comment:
Whoops, I didn't think through this before! `toString` should not have side effects. Could you do this in `setNewValue` instead and maybe change the name to `setNewValueOrClearIfUnchanged`? That's a mouthful, maybe `recordValueChange` or `saveValueChange`?

Reply:
Done. No worries I'm patient :wink:

Reply:
🙏 🙏